### PR TITLE
[FLINK-10689][table] Port UDFs in Table API extension points to flink-table-common

### PR DESCRIPTION
**Pull request by xueyumusic**
## What is the purpose of the change

This PR tries to port UDFs interfaces of TableApi extension points to flink-table-common in java. Right now  it ports UserDefinedFunction, ScalarFunction, TableFunction, AggregateFunction, DistinctAggregateFunction in org.apache.flink.table.functions execpt FunctionService and TemporalTableFunction. FunctionService and TemporalTableFunction may be could be ported following.

## Brief change log

*(for example:)*
  - *port several scala file in org.apache.flink.table.functions to java *

## Verifying this change

This change is a code refactor with existed tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/AggregateFunction.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ */
+
+package org.apache.flink.table.functions;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+
+/**
+ * Base class for User-Defined Aggregates.
+ *
+ * <p>The behavior of an [[AggregateFunction]] can be defined by implementing a series of custom
```
[the remark this answers is not in the document]
Comment (xueyumusic) [reply flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/AggregateFunction.java]: Thanks~, @bowenli86 , updated.

**Comment by bowenli86** [comment]
LGTM. +1 to merge

File: flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/AggregateFunction.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ * <p>Processes the input values and update the provided accumulator instance. The method
+ * accumulate can be overloaded with different custom types and arguments. An AggregateFunction
+ * requires at least one accumulate() method.
+ *
+ * {@code
+ * param accumulator           the accumulator which contains the current aggregated results
+ * param userDefinedInputs the input value (usually obtained from a new arrived data).
+ *
+ * def accumulate(accumulator: ACC, userDefinedInputs): Unit
```
Comment (twalthr) [inline flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/AggregateFunction.java]: Remove Scala code in all comments.

File: flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/AggregateFunction.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+ * }
+ *
+ *
+ * @param T   the type of the aggregation result
+ * @param ACC the type of the aggregation accumulator. The accumulator is used to keep the
+ *             aggregated values which are needed to compute an aggregation result.
+ *             AggregateFunction represents its state using accumulator, thereby the state of the
+ *             AggregateFunction must be put into the accumulator.
+ */
+public abstract class AggregateFunction<T, ACC> extends UserDefinedFunction {
```
Comment (twalthr) [inline flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/AggregateFunction.java]: Add proper annotations to all classes.

File: flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/DistinctAggregateFunction.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions;
+
+/**
+ * Defines an implicit conversion method (distinct) that converts {@link AggregateFunction}s into
+ * {@link DistinctAgg} Expressions.
+ */
+public class DistinctAggregateFunction<T, ACC> {
```
Comment (twalthr) [inline flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/DistinctAggregateFunction.java]: This class is not public API and should not be part of {{flink-table-common}}.

**Pull request closed without merging**